cpp/do_module/do_module.cpp

```diff
@@ -0,0 +1,313 @@
+                                                    "STORAGE MODE IN_MEMORY_ANALYTICAL"};
+
+struct ParamNames {
+  std::vector<std::string> node_names;
+  std::vector<std::string> relationship_names;
+  std::vector<std::string> primitive_names;
+};
+
+struct QueryResults {
+  std::vector<std::string> columns;
+  std::vector<std::vector<mg::Value>> results;
+};
+
+ParamNames ExtractParamNames(const mgp::Map &parameters) {
+  std::vector<std::string> node_names;
+  std::vector<std::string> relationship_names;
+  std::vector<std::string> primitive_names;
+
+  for (const auto &map_item : parameters) {
+    if (map_item.value.IsNode()) {
+      node_names.push_back(std::string(map_item.key));
+    } else if (map_item.value.IsRelationship()) {
+      relationship_names.push_back(std::string(map_item.key));
+    } else {
+      primitive_names.push_back(std::string(map_item.key));
+    }
+  }
+
+  return ParamNames{
+      .node_names = node_names, .relationship_names = relationship_names, .primitive_names = primitive_names};
+}
+
+std::string ConstructQueryPreffix(const ParamNames &names) {
+  if (!names.node_names.size() && !names.relationship_names.size() && !names.primitive_names.size()) {
+    return "";
+  }
+
+  std::vector<std::string> with_entity_vector;
+  for (const auto &node_name : names.node_names) {
+    with_entity_vector.push_back(fmt::format("${} AS __{}_id", node_name, node_name));
+  }
+  for (const auto &rel_name : names.relationship_names) {
+    with_entity_vector.push_back(fmt::format("${} AS __{}_id", rel_name, rel_name));
+  }
+  for (const auto &prim_name : names.primitive_names) {
+    with_entity_vector.push_back(fmt::format("${} AS {}", prim_name, prim_name));
+  }
+
+  auto with_variables = fmt::format("WITH {}", boost::algorithm::join(with_entity_vector, ", "));
+
+  std::string match_string = "";
+  std::vector<std::string> match_by_id_vector;
+  for (const auto &node_name : names.node_names) {
+    match_by_id_vector.push_back(fmt::format("MATCH ({}) WHERE ID({}) = __{}_id", node_name, node_name, node_name));
+  }
+  for (const auto &rel_name : names.relationship_names) {
+    match_by_id_vector.push_back(fmt::format("MATCH ()-[{}]->() WHERE ID({}) = __{}_id", rel_name, rel_name, rel_name));
+  }
+
+  if (match_by_id_vector.size()) {
+    match_string = boost::algorithm::join(match_by_id_vector, " ");
+  }
+
+  return fmt::format("{} {}", with_variables, match_string);
+}
+
+std::string ConstructPreffixQuery(const mgp::Map &parameters) {
+  const auto param_names = ExtractParamNames(parameters);
+
+  return ConstructQueryPreffix(param_names);
+}
+
+std::string ConstructFinalQuery(const std::string &running_query, const std::string &preffix_query) {
+  return fmt::format("{} {}", preffix_query, running_query);
+}
+
+mg::Map ConstructParams(const ParamNames &param_names, const mgp::Map &parameters) {
+  auto new_params = mg::Map(parameters.Size());
```

Comment:
I think this is better option as it avoids calling copy ctor:
 `mg::Map new_params{parameters.Size()};`